(BKR-1118) add puppet5 install method

kevpl opened the pull request:
![](http://www.dhresource.com/260x260s/f2-albu-g5-M00-01-C7-rBVaI1gm0a-ADYXnAAIDknLJ8AI322.jpg/5-pcs-sets-of-christmas-plush-toys-baby-five.jpg)

**Note** that this does not plug into `install_puppet` or `install_puppet_agent_on` methods at this point, it only provides basic puppet5 install functionality using the new `puppet5_install` method.

**Note** that that doesn't need to block merging this work, as those values need outside information that we don't have right now & can get in in a subsequent PR.

glennsarti commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,109 @@
+      # @param [String] project_sha SHA ref of the project to install
+      #
+      # @return nil
+      def puppet5_install(project_name, project_sha)
```
Does this mean we need to change our helper invocations to `puppet5_install` in order to use puppet 5 instead of puppet 4?

glennsarti replied:
Oh nvm...read the PR description

joshcooper commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,109 @@
+          platform_key.sub!('x86_64', 'amd64')
+        elsif variant == 'fedora'
+          platform_key.sub!(version, "f#{version}")
+        end
```
Can we do this translation in packaging instead of beaker, so that beaker stays dumb?

kevpl replied:
@joshcooper I don't see why not, but when we originally got together with @mwaggett, I told her to just use her normal packaging platform keys, since it's arbitrary which side of the packaging-beaker conversation figures them out.

That being said, this info is all OS-specific variation, which makes me think that the packaging repo would be a better place to handle it. Now that we've figured it out though, it seems like we could take our time moving that over.

@mwaggett does this make sense?

mwaggett replied:
@kevpl It kind of doesn't make sense to me for packaging to need to know how your test hosts are named, but I am open to discussing this. That being said, we do have a [ticket](https://tickets.puppetlabs.com/browse/RE-8699) for unifying platform naming (though we probably won't get to that for a while). 

I think the issue is that whenever one of us changes a platform tag (e.g. we're about to drop the 'f' from fedora versions), it will require a change in the other place. Another thing that will be solved by the metadata service, when we get to it..

kevpl replied:
@mwaggett these aren't dealing with details of how test hosts are named, but with inconsistencies in the keys provided in the `platform_data` data structure.

Glad to hear that fedora's f-version prefix will be going away soon. Packaging provides two different 64-bit architecture strings: `x86_64` or `amd64`. I assume that this distinction means something for packaging? For beaker, it doesn't care between them & so we've consistently named them `x86_64`. Even when we move to the metadata service, the problem will exist until we agree on what a consistent naming standard.

mwaggett replied:
@kevpl We use `amd64` for debuntus, I think to be consistent with how they do it upstream.

How strongly do you and @joshcooper feel about moving that logic to packaging? I'm happy to have a conversation about it, but I kind of don't see a huge benefit.

kevpl replied:
@mwaggett I'm not sure how @joshcooper feels about it, but as I'm seeing it, the question is really about the tradeoffs between either:

- being consistent across the OSes
- being consistent internally to Puppet's tooling

If we're consistent across the OS vendors, then we have to make sure that OS-specific changes are propagated across all of our tooling (which means across teams). Then, when those details change (& we can't control when that might happen), we have to update all of our projects to fit.

If we're consistent across internal Puppet tooling, then packaging becomes the place where the mapping between OS-specific details & puppet tooling takes place. Packaging would be the [adapter](https://en.wikipedia.org/wiki/Adapter_pattern) between Puppet (the company) & OS vendors. Then if something changes, the adapter can be updated by our OS-specific experts internally, & no one else has to miss a beat.

Note that we're off on a tangent of future work here. This doesn't need to block this PR merging, as pulling this case statement out would be trivial to do in the future, the only concern being a little scheduling between beaker-puppet & packaging work changes.

mwaggett replied:
yeah, I agree that internal consistency with packaging as an interface seems most desirable, but that is future work. Mostly I'm just eager to get this stuff in now. 😄

joshcooper replied:
Sorry to be the downer at the party, but don't we need to fix the naming issue before making these changes? For example, take the fedora case. In beaker+packaging v1, we have an acceptance job with `TEST_TARGET=fedora25-64a`. It's run through `beaker-hostgenerator` to generate a host config like:

```yaml
---
HOSTS:
  fedora25-64-1:
    hypervisor: vmpooler
    platform: fedora-25-x86_64
    template: fedora-25-x86_64
    roles:
    - agent
```

Beaker uses `platform=fedora-25-x86_64` internally. With this PR, we magically change it to `fedora-f25-x86_64`. We then use that key to lookup the build artifact that the packaging repo created.

In the future, packaging v2 will no longer include `f` in the key. Now we need to update beaker to match. But beaker needs to be able to install packages created by v1 and v2 of packaging, e.g. in the puppet-agent LTS pipeline (1.10.x) and the latest 5.x pipeline. Since the beaker change will break older pipelines, we have to take a semver major, or add a feature flag that consumers can opt-into, or add logic to try `f25` and fallback to `25`. In other words, we find ourselves reimplementing the same set of hacks in beaker, that we're trying to avoid in the beaker+packaging epic.

In my mind, beaker should be data-driven. So given a `key`, lookup the packaging artifact/repo in the hash produced by the packaging repo. So long as the key doesn't change for a given OS/arch combo, then we can make changes in the packaging layout without breaking any consumer, e.g. `fc25`. But because this is a bootstrapping problem, we have to get the name right to begin with. I don't think the key necessarily needs to be `platform`, eg it could be something like `host[:packaging_name]`, etc

mwaggett replied:
@joshcooper @kevpl so if I'm understanding correctly, you would like the fedora platform tags to always be of the form fedora-<ver>-<arch> and to always use x86_64 instead of a mix of that and amd64? The former should be easy; I will look into what it would take for the latter.

joshcooper commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,109 @@
+          on host, "wget -O #{ onhost_package_file } #{ artifact_url }"
+        elsif variant == 'osx'
+          on host, "curl -O #{ artifact_url }"
+          onhost_package_file = "#{ project_name }*"
```
Do beaker's `Host` classes already have a method for downloading/fetching a file, e.g. `host.download_from_url(artifact_url)`, so that we don't need if/then/else, and knowledge about wget vs curl?

kevpl replied:
hm, I'm not seeing a method that abstracts between these two right off the bat.

There's a `fetch_http_file` method that we often use with `scp_to` to install PE artifacts ([`install_repo_configs` example](https://github.com/puppetlabs/beaker-puppet/blob/master/lib/beaker-puppet/install_utils/foss_utils.rb#L966-L975)), but if we can get them right to the SUT rather than having to go through the coordinator that would be more efficient.

This pattern is pretty typical in the FOSS installs (just straight using `on` to execute curl or wget), but we don't typically see it side-by-side because of the way the work is split up in the install code (historically, we've done _everything_ for a platform in one body of a switch statement, rather than break up the tasks of download artifact vs install artifact). Refactoring this work into a helper method seems like a useful thing, but something that can be done separately from this work.

@joshcooper am I just missing the method, or do you think creating that helper method should block this PR?

joshcooper replied:
Ideally, we shouldn't open-code the logic for what it means to execute curl/wget across all platforms, but that's always been an issue in beaker. So I don't think it should block this PR.

tvpartytonight commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,109 @@
+      # @return [String, String] URL to the build for the SHA of the project,
+      #   build json parsed into a ruby hash
+      def fetch_build_details(project_name, project_sha)
+        buildserver_url = 'http://builds.delivery.puppetlabs.net'
```
Does this need to be configurable for users outside of the puppet firewall?

tvpartytonight commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,109 @@
+        buildserver_projectsha_url = "#{buildserver_url}/#{project_name}/#{project_sha}"
+        file_folder = "#{buildserver_projectsha_url}/artifacts"
+        filename = "#{project_sha}.yaml"
+        dst_folder = Dir.pwd
```
Is there a reason we put this in the local directory, as opposed to using `Dir.mktmpdir`?

joshcooper replied:
👍  to use `mktmpdir`, 1) because it's a secure directory, and 2) because it will be unique for each run, so any build artifacts from a previous run, e.g. `puppet-agent-latest-x64.msi`, won't pollute future runs.

tvpartytonight commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,109 @@
+        file_hash = YAML.load_file(file_local_path)
+
+        # getting the relevant data from the file
+        logger.perf_output("YAML HASH BELOW:")
```
Why are these `perf_output` instead of `debug` or `info`?

kevpl commented:
@tvpartytonight thanks for the 👀 , I think most of those were just me not cleaning that particular method up for the PR, so thanks for the catch!

@joshcooper I had a wrong idea of how versions would work across both repos, so thanks for bringing up that situation! I like the data-driven bits & how we can get around dealing with the changes from the beaker side using a different property than platform-munging, so I've gone that route in the latest update! Let me know if there are any parts of that objection that still stand, or if you have any feedback on the new changes.

@mwaggett, this should make beaker work with changes across the packaging repo, as we've moved dealing with that to the host's configuration rather than dealing with it in beaker code. I'm glad we had the conversation about naming since it'll be important when we move to reconcile these though, so thanks for your help!

kevpl commented:
@joshcooper @mwaggett I've updated this PR with the changes we talked about from Friday: making `puppet5_install` take a full URL rather than parameters & building the URL out of it.

Note that my local testing fails, but that's because of the relative to the <SHA>.yaml file work that hasn't come through to the builds server that I'm seeing. I'll give an example of this in the BKR ticket, so we can be confident that this behaves correctly.

mwaggett commented:
@kevpl These packages were built with the changes we discussed: http://builds.delivery.puppetlabs.net/puppet-agent/ff3e02ddf16781f94b50918b995c6e793597f9a9/
(https://github.com/puppetlabs/packaging/pull/690)

kevpl commented:
@mwaggett thank for that ref! I'm able to test it & see that it goes green with my newest code from yesterday.

Forgot to change the name from `packaging_name` to `packaging_platform` yesterday, so I did that this morning. @mwaggett @joshcooper this is ready for review

mwaggett commented:
since I'm not familiar with this codebase, I don't know how valuable a review from me is, but if you're able to fetch and test the builds I made with my PR, then this looks good to me!

joshcooper commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,128 @@
+        if onhost_package_file
+          host.install_local_package( onhost_package_file )
+        else
+          host.install_package( artifact_url )
```
Trying this on ubuntu-16.04-amd64, this method ends up calling:

```
q2p1dby68kwqdaq.delivery.puppetlabs.net (q2p1dby68kwqdaq.delivery.puppetlabs.net) 14:06:34$ apt-get install --force-yes  -y http://builds.delivery.puppetlabs.net/puppet-agent/f0005a296b2e1f5d6a7541aac005d41550690bc5/artifacts/./deb/xenial/PC1/puppet-agent_5.0.0.137.gf0005a2-1xenial_amd64.deb
    Reading package lists...
    Building dependency tree...
    Reading state information...
    E: Unable to locate package http://builds.delivery.puppetlabs.net/puppet-agent/f0005a296b2e1f5d6a7541aac005d41550690bc5/artifacts/./deb/xenial/PC1
    E: Couldn't find any package by glob 'http://builds.delivery.puppetlabs.net/puppet-agent/f0005a296b2e1f5d6a7541aac005d41550690bc5/artifacts/./deb/xenial/PC1'
    E: Couldn't find any package by regex 'http://builds.delivery.puppetlabs.net/puppet-agent/f0005a296b2e1f5d6a7541aac005d41550690bc5/artifacts/./deb/xenial/PC1'
```

But it fails because `apt-get` can't seem to install the `.deb` from a URL. Should we curl/wget the `deb` locally to the SUT first?

kevpl replied:
@joshcooper added ubuntu to the local install case. Which job should I be looking to to complete the platform array for this install method?

joshcooper replied:
These are the current list of BHG platforms we test against: https://github.com/puppetlabs/ci-job-configs/blob/626f1617cca7814c0b01299998573aa949e3f67e/jenkii/jenkins-master-prod-1/projects/puppet-agent.yaml#L207-L231

kevpl replied:
@joshcooper working on testing these platforms, @melissa noticed that ubuntu1610 was on the list & thought we weren't shipping puppet-agent on that platform right now. Checking with CI peeps, it sounds like the up-to-date version of the platform array would just be the file on the [master branch](https://github.com/puppetlabs/ci-job-configs/blob/master/jenkii/jenkins-master-prod-1/projects/puppet-agent.yaml#L157-L180) for those pipelines, but it looks like that has ubuntu1610 as well. What would be the right version of that file that we should be testing from to get the correct array of platforms?

mwaggett replied:
@kevpl ubuntu1610 has been EOL'd for puppet5 and beyond, but is currently still being shipped in the 1.10.x series. I think the plan is to EOL it there too, but that hasn't been done yet.

mwaggett replied:
https://github.com/puppetlabs/build-data/blob/release/builder_data.yaml#L64 is a pretty solid list (and I think currently accurate), but sometimes we forget to remove EOL'd things.

kevpl replied:
thanks for clearing this up @mwaggett! 🥇

melissa replied:
Hey all! We're trying to have a single source of truth for the supported platforms. Anything in github should be based on the confluence page, but should not be considered a final source of truth. For the final source of truth, consult https://confluence.puppetlabs.com/display/PROD/Currently+Supported+OS+Platforms. However, it does look like it has out of date information for yakkety :(

A lot of this decision not to ship Yakkey comes from the following email conversation between Ruth, @joshcooper and @haus 

>>> (Note: probably only the platform team cares about this, but I don't know what the appropriate email list is for that that definitely includes the people who care about this and I already spent 5 minutes trying to figure it out before giving up.)

>>>The plaftorm team has done some work to get puppet-agent and puppet server packages for Yakkety (Ubuntu 16.10) for Puppet 5... but there are still bugs to sort out causing redness in pipelines.

>>Curious if the bugs are specific to Yakkety-only, or will we hit the same issues as soon as we try to add support for Zesty 17.04 in https://tickets.puppetlabs.com/browse/PA-873?

>Yakkety has the weird puppet-agent package that isn't from us, while Zesty does not, fwiw.
 
>>>Given that the current date for GA for Puppet 5 is end of June and Yakkety goes EOL in July, it seems like rather than trying to spend more time getting this to work, our time could be better spent elsewhere.

>>>I talked with Eric0 briefly and he was on board with not shipping Yakkety packages for Puppet platform 5, but I wasn't sure what further things (confluence pages??) need updating, and also wanted to make sure this discussion was widely circulated.

joshcooper commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,128 @@
+      # @param [String] project_name Name of the project to install
+      #
+      # @return nil
+      def puppet5_install(project_name, sha_yaml_url)
```
If this method should be usable for projects like `puppetserver`/`puppetdb` that have dependencies, then I think we always want to install the `repo_config` for platforms with package managers, and only install the package directly for things like msi/dmg/etc.

Also I'm thinking this isn't puppet5 specific, more like `install_from_build_data_url`?

kevpl commented:
@joshcooper I've added `repo_config` installation as a step.

I've changed the `puppet_install` name as well & added a comment to note that it was made in puppet 5 for deprecation purposes later.

joshcooper commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,162 @@
+          scp_to host, artifact_filename, '/tmp'
+          onhost_package_file = "/tmp/#{ artifact_filename }"
+        elsif variant =~ /^(debian|ubuntu)$/
+          onhost_package_file = '/tmp/puppet.deb'
```
Can we use `host.tmpfile('puppet.deb')` to ensure we don't have conflicts?

melissa commented:
Hey all! Any movement on this?

kevpl commented:
Hey @melissa, I'm working on getting the platform coverage that we're looking for, but I'm hitting some difficulties testing the different platforms with the vmpooler degredation we're experiencing, so that's slowing this effort down. I'll be keeping an eye on that, & coming back to it as soon as I can (hopefully in the next day or so).

melissa commented:
Thank you so much for the update, @kevpl ! Let me know if I can be of any help

kevpl commented:
@joshcooper @melissa updated w/`tmpfile` improvements & additional platform testing. `eos` was the only platform that needed additional logic, & some of this will be improved after this work is released. Apologies with the delay, there was a snag dealing with some `cisco` testing, but now we're through that. Let me know if you have any questions or comments about this.

melissa commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,176 @@
+        hosts.each do |host|
+          artifact_url, repoconfig_url = host_urls( host, build_details, base_url )
+          install_repo_configs_on( host, repoconfig_url )
+          install_artifact_on( host, artifact_url, project_name )
```
This kind of looks like we're installing the repo_configs if they're available, but then not using those enabled repos? In `install_artifact_on`, it looks like we're just installing the artifact directly from the `artifact` url. Am I misinterpreting this?

kevpl commented:
@melissa I've pushed new changes that take advantage of the repo_config info if it exists. Let me know if you see any other issues with this. Thank you!

melissa commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,177 @@
+            host.install_local_package( onhost_package_file )
+          end
+        else
+          host.install_package( artifact_url )
```
So this will give the whole artifact url as defined in the yaml file, yes? Like `http://builds.puppetlabs.lan/puppet-agent/1.9.1/artifacts/el/7/PC1/x86_64/puppet-agent-1.9.1-1.el7.x86_64.rpm` or `http://builds.puppetlabs.lan/puppet-agent/903bd6cbb8bfc6f88a6be61acc86b8bfa12b8690/artifacts/deb/xenial/PC1/puppet-agent_5.0.1.113.g903bd6c-1xenial_amd64.deb`?

I'm not 100% sure all package management sysmtes are going to be able to handle that as an argument, unless it's getting parsed somewhere that I've missed

```
root@plb4iyu3cgus7pk:~# apt-get install --force-yes -y http://builds.puppetlabs.lan/puppet-agent/903bd6cbb8bfc6f88a6be61acc86b8bfa12b8690/artifacts/deb/xenial/PC1/puppet-agent_5.0.1.113.g903bd6c-1xenial_amd64.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package http://builds.puppetlabs.lan/puppet-agent/903bd6cbb8bfc6f88a6be61acc86b8bfa12b8690/artifacts/deb/xenial/PC1
E: Couldn't find any package by glob 'http://builds.puppetlabs.lan/puppet-agent/903bd6cbb8bfc6f88a6be61acc86b8bfa12b8690/artifacts/deb/xenial/PC1'
E: Couldn't find any package by regex 'http://builds.puppetlabs.lan/puppet-agent/903bd6cbb8bfc6f88a6be61acc86b8bfa12b8690/artifacts/deb/xenial/PC1'
```

melissa replied:
Oh, wait, this might not matter, right? So, this method will only get called if repo_configs is nil, so things without a repo (osx, eos, etc) are the only platforms that should be using this method unless for some reason things broke and those repo_configs weren't created...

kevpl replied:
@melissa yes, this code path will only get called with full artifact URLs when repo_configs aren't set.

In terms of your error question, one of the goals of these changes is that we want beaker to remain dumb & not able to know when repo_configs should & shouldn't be available, so testing for this would be difficult unless we broke that. It seems like the packaging system that generates the repo_configs would be better suited to validating repo_config existence before saying a build is finished.

I can see how validating & behaving nicely on both sides would be a nice experience for the end user, but I think that's how we often end up with poor duplicate copies between beaker & the packaging system. 

Right now the [output of a failure case](https://github.com/puppetlabs/beaker/blob/master/lib/beaker/dsl/helpers/web_helpers.rb#L63-L67) is pretty explicit about what caused it to fail if these are unreachable, so I don't think it's hard to debug. There are some other options for dealing with failure cases, but I think we'll talk about them in the other comment that's on the specific lines in question.

melissa commented on lib/beaker-puppet/install_utils/puppet5.rb:
```diff
@@ -0,0 +1,177 @@
+            install_artifact_on( host, artifact_url, project_name )
+          else
+            install_repo_configs_on( host, repoconfig_url )
+            host.install_package( project_name )
```
💯

melissa commented on lib/beaker-puppet/install_utils/foss_utils.rb:
```diff
@@ -963,9 +963,27 @@ def install_repo_configs(host, buildserver_url, package_name, build_version, cop
+          repoconfig_filename = File.basename(  repo_config_url )
+          repoconfig_folder   = File.dirname(   repo_config_url )
+
+          repo = fetch_http_file(
```
Is there a graceful way to fail back to installing the package directly if for some reason something broke and the repo_config files aren't available? That actually might not be something we want to allow beaker to do, now that I'm thinking about it....

kevpl replied:
@melissa I'm wary of degrading to a different install if the repo_configs aren't available. That's something that we've done historically, and we've had two problems with that approach: transparency & debugging/maintenance.

One is that magic changes underneath a user (like an auto install-method switch) is much harder to debug & understand. The only way to be transparent about degrading in this fashion that we've seen is to scream into the console log, but then we face the problem of screaming wolf too many times & nobody pays attention to the log colors/warnings/errors anymore.

Another is that the more avenues we use to install, the harder it is to maintain that code. If the issue is potentially in the packaging system such that the symptom is that repo_config files aren't being made available, then finding alternate ways to install creates more distance between the cause of errors & the way that we find them. It seems that failing at build-time would be even better, but failing as early as possible during the install step is the best that beaker can do to help a user find a bug like that. 

If we're thinking of transient issues, like network blips over getting the appropriate file rather than the file being in the right place, then there are some things we can definitely do to help with that. Our usual suggestion at this point would be to back off & retry, hoping that those network issues resolve themselves over a small break. Unfortunately, this can hide issues & delay job completion, but it can help overcome things like this. I would say at this point that there's no reason to add that complexity if we don't need it though.

melissa replied:
I agree those are good reasons to be wary of that course of action. I think we should move forward with this as is

melissa replied:
A little followup, yes, the error message here when this file is missing is very informative. But here's a question I should have asked earlier: why not derive the repo_config info from the yaml file? Right now we're guessing what the repo name should be, which could be wrong
```
           repo_filename = host.repo_filename( package_name, build_version )
            repo_config_folder_url = "%s/%s/%s/repo_configs/%s/" %
              [ buildserver_url, package_name, build_version, host.repo_type ]
```

melissa replied:
hmm, nope, this is totally as it should be. The new functionality should be and is isolated. My bad. IGNORE ME.

melissa commented:
I think this looks great, and I can't wait to see it in action!

kevpl commented:
@joshcooper @tvpartytonight if you have some time, I'd appreciate another look at this.

melissa commented:
@joshcooper @tvpartytonight I would love to get this in soon so we can start leveraging it for those projects targeted at puppet5!

melissa commented:
Also, we have a good opportunity to update the beaker version in the puppet agent 5.x branch tomorrow after the latest release. I'd love to get the tests going and suss out any potential issues we might run into

kevpl commented:
@puppetlabs/beaker it would be great to get some more eyes on this too!

melissa commented:
Hey @puppetlabs/beaker @joshcooper @tvpartytonight puppet agent 5.1.0 went out yesterday, so we have a great window to bump the beaker version in that branch!

melissa commented:
@kevpl it looks like there's a merge conflict that needs to be resolved :(

kevpl commented:
merge conflict resolved

kevpl commented:
The mirror didn't have oga from Rishi's PR today, so I pushed it up.

@puppetlabs-jenkins retest this please.

kevpl commented:
same with ruby-ll, ugh.

@puppetlabs-jenkins retest this please

tvpartytonight commented:
whoooooo!

melissa commented:
YAY!
